### ISSUE TITLE
List comprehensions and for-loops on xla.DeviceArrays return numpy.ndarrays

### ISSUE THREAD

**Issue by jessebett** [google/jax#1583]
Looping through a `DeviceArray` like in a list comprehension, for loop, or enumerate returns `numpy.ndarray`. Is this expected behaviour?

```python
import jax.numpy as np

X = np.array(range(3)).reshape(3,1)
print(type(X))
#<class 'jax.interpreters.xla.DeviceArray'>

x0,x1,x2 = [xi for xi in X]
print(type(x0))
#<type 'numpy.ndarray'>

for xi in X:
  print(type(xi))
#<type 'numpy.ndarray'>
#<type 'numpy.ndarray'>
#<type 'numpy.ndarray'>
```

This is unexpected to me. Why are these not also `DeviceArray`s?

**Comment by mattjj** [comment]
Thanks for the question! It is intended behavior: I used git blame to pull up commit 71605f4, which implemented this behavior by changing the implementation of `DeviceArray.__iter__` in xla.py. The commit message has more information:

> Change `DeviceArray.__iter__` and `DeviceArray.__reversed__` to forward to the _value.
>
> This has the effect of transferring the entire array to the host and iterating over it in host memory, rather than slicing out individual elements in device memory one by one.
> 
> This is much faster for examples like `list(np.arange(10000))`; previously this took several seconds the first time due to compilation and 100ms+ subsequent times. With this change it takes < 1ms.

In short, we do it for performance reasons.

In general, JAX doesn't guarantee when it will return an ndarray versus a DeviceArray. That lets us adjust the behavior in cases like this one.

WDYT?

**Comment by jessebett** [comment]
Okay, thanks @mattjj. I was pretty sure that I don't want to be looping through these `DeviceArrays` anyway, but now am confident that I'll need to do a vectorized approach. Correct me if I'm wrong but the drop down to ndarray also drops gradient tracking?

**Comment by jessebett** [comment]
Actually, @mattjj this shows up in more surprising ways that I think are a problem for me. 

For the `jet` stuff, we want to have jet coefficents as a list of `DeviceArrays`. However, `prop` rules for many of the primitives are more efficient if you treat them together as a multidimensional `DeviceArray`. So not changing the list-of-coefficients API, internally I concatenate them into a multidimensional array, perform the propogation rules, which produces a new multidimensional array of the output coefficients. Now, I'd expect that I could just call `list` on this to return the list of `DeviceArrays`, but this is not supported:

```python
In [7]: import jax.numpy as np                                                                              

In [8]: np.array(np.arange(10)).reshape(5,2)                                                                
Out[8]: 
DeviceArray([[0, 1],
             [2, 3],
             [4, 5],
             [6, 7],
             [8, 9]], dtype=int32)

In [9]: list(np.array(np.arange(10)).reshape(5,2))                                                          
Out[9]: 
[array([0, 1], dtype=int32),
 array([2, 3], dtype=int32),
 array([4, 5], dtype=int32),
 array([6, 7], dtype=int32),
 array([8, 9], dtype=int32)]
```

Possibly there's a workaround or a different approach that is more appropriate in general.

**Comment by jekbradbury** [comment]
Why can't you use multidimensional arrays throughout as the representation of jet coefficients?

**Comment by jessebett** [comment]
The performance gains wasn't clear (to me) until recently. So another solution might look like rewriting our assumptions on the types of the coefficents. This would actually be non-trivial, though. I believe I had them as a multidimensional array at one point. @mattjj do you remember if we had a good reason to write the coefficients as lists of primal-dimensional arrays instead of together as a order-dimensional array?
